blockEntity.js

```diff
@@ -3,66 +3,101 @@ const nbt = require('prismarine-nbt')
 module.exports = registry => {
   if (registry.version.type === 'pc') {
     const ChatMessage = require('prismarine-chat')(registry.version.majorVersion)
-    return {
-      sign: {
-        get blockEntity () {
-          // Compatibility for mineflayer, which changes .blockEntity for signs to contain ChatMessages instead of simplified NBT
-          if (!this.entity) return undefined
 
-          const prepareJson = (i) => {
-            const data = this.entity.value[`Text${i}`].value
-            if (!data || data === '') return ''
-            const json = JSON.parse(data)
-            if (json === null || !('text' in json)) return ''
-            return json
+    function setSignTextForMultiSideSign (block, side, text) {
+      const texts = []
+      if (typeof text === 'string') {
+        // Sign line should look like JSON string of `{"text: "actualText"}`. Since we have plaintext, need to add in this JSON wrapper.
+        texts.push(JSON.stringify(text.split('\n').map((t) => ({ text: t }))))
+      } else if (Array.isArray(text)) {
+        for (const t of text) {
+          if (t.toJSON) { // prismarine-chat
+            texts.push(JSON.stringify(t.toJSON()))
+          } else if (typeof t === 'object') { // normal JS object
+            texts.push(JSON.stringify(t))
+          } else { // plaintext
+            texts.push(JSON.stringify({ text: t }))
           }
+        }
+      }
 
-          return {
-            id: this.entity.value.id || 'minecraft:sign',
-            Text1: new ChatMessage(prepareJson(1)),
-            Text2: new ChatMessage(prepareJson(2)),
-            Text3: new ChatMessage(prepareJson(3)),
-            Text4: new ChatMessage(prepareJson(4))
-          }
-        },
+      if (!block.entity) {
+        block.entity = nbt.comp({
+          id: nbt.string(registry.version['>=']('1.11') ? 'minecraft:sign' : 'Sign')
+        })
+      }
 
-        set signText (text) {
-          const texts = []
-          if (typeof text === 'string') {
-            // Sign line should look like JSON string of `{"text: "actualText"}`. Since we have plaintext, need to add in this JSON wrapper.
-            texts.push(JSON.stringify(text.split('\n').map((t) => ({ text: t }))))
-          } else if (Array.isArray(text)) {
-            for (const t of text) {
-              if (t.toJSON) { // prismarine-chat
-                texts.push(JSON.stringify(t.toJSON()))
-              } else if (typeof t === 'object') { // normal JS object
-                texts.push(JSON.stringify(t))
-              } else { // plaintext
-                texts.push(JSON.stringify({ text: t }))
-              }
-            }
+      block.entity.value[side].value.messages.value.value = texts
```

Comment:
Where are you generating the NBT tree here? The code just assumes it exists, which means it'll throw if a new sign empty is created. After entering line 25 for example, this will throw.